### PR TITLE
Models (state container & new JavaScript best practices) (WIP)

### PR DESCRIPTION
**Pull request by sebbert**
Replaces #377 (rebased onto current master)

This also contains #499

This PR contains:
- [x] Changelog
- [ ] Documentation
- [x] Tests

### CODE REVIEW

File: Source/Fuse.Controls.Navigation/NavigationControl.Pages.uno [hunk 1]
```diff
@@ -40,7 +40,7 @@ namespace Fuse.Controls
 			if (_pageHistory == null)
 				return;
 				
-			var obs = _pages as IObservableArray;
+			var obs = _pageHistory as IObservableArray;
```
Comment (kusma) [inline Source/Fuse.Controls.Navigation/NavigationControl.Pages.uno]: Could you squash this into the commit the mistake originated from?

File: CHANGELOG.md [hunk 2]
```diff
@@ -131,7 +131,6 @@ which will stop push notifications registering (and potentially asking for permi
 - Introduced support for variable arguments to UX functions - inherit from the `Fuse.Reactive.VarArgFunction` class.
 - The classes `Vector2`, `Vector3` and `Vector4` in `Fuse.Reactive` are now removed and replaced with the general purpose, variable-argument version `Vector` instead. This ensures vectors of any length are treated the same way. This is backwards incompatible in the unlikely case of having used these classes explicitly from Uno code.
 - Added support for name-value pair syntax: `name: value`. Can be used for JSON-like object notation and named arguments in custom functions. Any vector of name-value pairs is interpreted as an `IObject`, e.g. `{name: 'Joe', apples: 10}` is an object.
-- Added support for name-value pair syntax: `name: value`. Can be used for JSON-like object notation and named arguments in custom functions.
```
Comment (kusma) [inline CHANGELOG.md]: Please squash this into the commit it originated from.

File: Source/Fuse.Reactive.JavaScript/FuseJS/Model.js [hunk 3]
```diff
@@ -171,15 +171,12 @@ function Model(source)
...
 			isDirty = true;
-			console.log("setting timeout")
 			setTimeout(meta.diff, 0);
-			console.log("done setting timeout")
 		}
 
 		var changesDetected = 0;
 
 		meta.diff = function() {
-			console.log("Diffing!")
```
Comment (kusma) [inline Source/Fuse.Reactive.JavaScript/FuseJS/Model.js]: Please squash this into the commit it originated from.

File: Source/Fuse.Reactive.JavaScript/FuseJS/Model.js [hunk 4]
```diff
@@ -1,4 +1,9 @@
 var TreeObservable = require("FuseJS/TreeObservable")
+
+console.error = function(e) {
+	console.log(JSON.stringify(e))
+}
+
```
Comment (kusma) [inline Source/Fuse.Reactive.JavaScript/FuseJS/Model.js]: This bit seems irrelevant to the commit message.

**Comment by sebbert** [comment]
@kusma Indeed, I'll clean it up

**Comment by sebbert** [comment]
@kusma I've gone through and cleaned up the history quite a bit, but I'm not entirely sure how granular you want this to be. Let me know if you'd like i.e. bugfixes squashed as well.

File: Source/Fuse.Reactive.JavaScript/TreeObservable/TreeObservable.uno [hunk 5]
```diff
@@ -124,7 +127,7 @@ namespace Fuse.Reactive
 			void PerformStart()
 			{
 				Perform(TreeObservable, 0);
-			}
+			}
```
Comment (kusma) [inline Source/Fuse.Reactive.JavaScript/TreeObservable/TreeObservable.uno]: this space seems like a mistake

File: Source/Fuse.Reactive.JavaScript/JavaScript.Model.uno [hunk 6]
```diff
@@ -140,10 +140,7 @@ namespace Fuse.Reactive
                 {
 					var argName = "__dep" + i;
 					var c = nfc.Arguments[i] as Constant;
-                    if (nt != null && c != null && c.Value == nt.This)
-                    {
-						result.Dependencies.Add(new Dependency(argName, nfc.Arguments[i]));
-					}
+                    result.Dependencies.Add(new Dependency(argName, nfc.Arguments[i]));
```
Comment (kusma) [inline Source/Fuse.Reactive.JavaScript/JavaScript.Model.uno]: Bad indentation

**Comment by kusma** [comment]
Superseeded by #601
